src/cr/cube/matrix/cubemeasure.py

```diff
@@ -87,851 +94,1251 @@ def _slice_idx_expr(cls, cube, slice_idx):
+            else _MrXArrCubeCounts
+            if dimension_type_strings == ("MR", "ARR")
+            else _MrXCatCubeCounts
+            if dimension_type_strings == ("MR", "CAT")
+            else _ArrXMrCubeCounts
+            if dimension_type_strings == ("ARR", "MR")
+            else _ArrXArrCubeCounts
+            if dimension_type_strings == ("ARR", "ARR")
+            else _ArrXCatCubeCounts
+            if dimension_type_strings == ("ARR", "CAT")
+            else _CatXMrCubeCounts
+            if dimension_type_strings == ("CAT", "MR")
+            else _CatXArrCubeCounts
+            if dimension_type_strings == ("CAT", "ARR")
+            else _CatXCatCubeCounts
+        )
+
+        return CubeCountsCls(
+            dimensions, counts[cls._slice_idx_expr(cube, slice_idx)], diff_nans
         )
 
     @lazyproperty
-    def means(self):
-        """2D np.float64 ndarray of cube means."""
+    def column_bases(self):
+        """2D np.float64 ndarray of column-wise bases for each valid matrix cell."""
         raise NotImplementedError(  # pragma: no cover
-            "`%s` must implement `.means`" % type(self).__name__
+            "`%s` must implement `.column_bases`" % type(self).__name__
         )
 
+    @lazyproperty
+    def columns_base(self):
+        """Optional 1D np.float64 ndarray of column-wise base for each valid column.
 
-class _CatXCatCubeMeans(_BaseCubeMeans):
-    """Means cube-measure for a slice with no MR dimensions."""
+        None if rows dimension is an ARRAY or MR because the base can vary among columns
+        """
+        return None
 
     @lazyproperty
-    def means(self):
-        """2D np.float64 ndarray of means for each valid matrix cell."""
-        return self._means
+    def columns_table_base(self):
+        """Optional 1D np.float64 ndarray of table-wise base for each valid column.
 
+        The name is a mouthful, but each component is meaningful.
+        - "columns": Indicates it is a marginal in the "columns" orientation (kind of
+        like a stripe in the shape of a row).
+        - "table": Indicates that it is the base for the whole table. When the
+        `.table_base` exists (CAT X CAT), it is a repetition of that, but when
+        the rows are array (and therefore we can't sum across them), each cell has
+        its own value.
+        - "base": Indicates that it is the base, not necessarily the counts (eg the sum
+        of selected and non-selected for MR variables)
 
-class _CatXMrCubeMeans(_BaseCubeMeans):
-    """Means cube-measure for a NOT_MR_X_MR slice.
+        None if rows dimension is an ARRAY or MR because the base can vary among columns.
+        """
+        return None
 
-    Note that the rows-dimensions need not actually be CAT.
-    """
+    @lazyproperty
+    def counts(self):
+        """2D np.float64 ndarray of count for each valid matrix cell.
+
+        A valid matrix cell is one whose row and column elements are both non-missing.
```

Comment:
I think the "valid" bits and definition can go now that unconditional counts are conceptually far away.